File: app/src/main/java/antoninovitale/dropcodechallenge/list/model/mapper/BeerListModelMapper.java
package antoninovitale.dropcodechallenge.list.model.mapper;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import antoninovitale.dropcodechallenge.api.model.Beer;
import antoninovitale.dropcodechallenge.list.model.Attribute;
import antoninovitale.dropcodechallenge.list.model.BeerListModel;
import antoninovitale.dropcodechallenge.list.model.IBeerListModel;
import antoninovitale.dropcodechallenge.util.Utils;

/**
 * Created by antoninovitale on 31/08/2017.
 */
public class BeerListModelMapper {

    public static List<IBeerListModel> convertBeers(List<Beer> beers) {
        List<IBeerListModel> models = Collections.emptyList();
        if (beers != null && !beers.isEmpty()) {
            models = new ArrayList<>(beers.size());
            for (Beer beer : beers) {
                IBeerListModel model = convertBeer(beer);
                if (model != null) {
                    models.add(model);
                }
            }
        }
        return models;
    }

    private static IBeerListModel convertBeer(Beer beer) {
        return beer != null ? new BeerListModel(beer.getName(), beer.getTagline(), Utils
                .formatPercentage(beer.getAbv()), Attribute.getAttribute(beer.getAbv(), beer
                .getIbu()), beer.getImageUrl()) : null;
    }

}

File: app/src/main/java/com/ninovitale/punkapi/app/list/model/mapper/BeerListModelMapper.kt
package com.ninovitale.punkapi.app.list.model.mapper

import com.ninovitale.punkapi.app.api.model.Beer
import com.ninovitale.punkapi.app.list.model.Attribute
import com.ninovitale.punkapi.app.list.model.BeerListModel
import com.ninovitale.punkapi.app.list.model.IBeerListModel
import com.ninovitale.punkapi.app.util.Utils
import java.util.ArrayList

/**
 * Created by antoninovitale on 31/08/2017.
 */
object BeerListModelMapper {
    @JvmStatic
    fun convertBeers(beers: List<Beer?>?): List<IBeerListModel?> {
        var models: List<IBeerListModel?> = emptyList()
        if (beers != null && beers.isNotEmpty()) {
            models = ArrayList(beers.size)
            beers.mapNotNull { beer -> convertBeer(beer).also { models.add(it) } }
        }
        return models
    }

    private fun convertBeer(beer: Beer?): IBeerListModel? {
        return if (beer != null) BeerListModel(beer.name, beer.tagline,
                Utils.formatPercentage(beer.abv), beer.imageUrl,
                Attribute.getAttribute(beer.abv, beer.ibu)) else null
    }
}